services/src/main/java/com/linkedin/kmf/services/TopicRebalancer.java

```diff
@@ -0,0 +1,427 @@
+/**
+ * Copyright 2016 LinkedIn Corp. Licensed under the Apache License, Version 2.0 (the "License"); you may not use this
+ * file except in compliance with the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ */
+
+package com.linkedin.kmf.services;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import kafka.admin.AdminUtils;
+import kafka.admin.PreferredReplicaLeaderElectionCommand;
+import kafka.cluster.Broker;
+import kafka.common.TopicAndPartition;
+import kafka.utils.ZkUtils;
+import org.apache.kafka.common.Node;
+import org.apache.kafka.common.PartitionInfo;
+import org.apache.kafka.common.errors.InterruptException;
+import org.apache.kafka.common.security.JaasUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.linkedin.kmf.common.Utils.ZK_CONNECTION_TIMEOUT_MS;
+import static com.linkedin.kmf.common.Utils.ZK_SESSION_TIMEOUT_MS;
+
+import scala.collection.Seq;
+
+
+/**
+ * Runs this periodically to rebalance the monitored topic across brokers and to reassign leaders to brokers so that the
+ * monitored topic is sampling all the brokers evenly.
+ */
+class TopicRebalancer implements Runnable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(TopicRebalancer.class);
+
+  /**
+   * The state of the monitored topic.
+   */
+  enum RebalanceCondition {
```

Comment:
This kind of imply that these 4 conditions are mutually exclusive, but they are not. For example, we can have a case where a broker doesn't have any partition and thus satisfied the last 3 conditions.

And alternative is to have the following logic in run():
```
if (isPartitionLow) increasePartitionNum()
if (isAnyBrokerNotPreferredLeader()) reassignReplicas()
if (isAnyBrokerNotElectedLeader()) doPreferredLeaderElection()
```